nx/lib/nx/binary_backend.ex

```diff
@@ -427,7 +427,54 @@ defmodule Nx.BinaryBackend do
   end
 
   @impl true
-  def dot(out, %{type: t1} = left, axes1, %{type: t2} = right, axes2) do
+  def dot(out, left, axes1, nil, right, axes2, nil) do
+    dot4(out, left, axes1, right, axes2)
+  end
+
+  def dot(out, left, axes1, batch1, right, axes2, batch2) do
+    {left_tensors, axes1, s1, names1} = prepare_batch(left, axes1, batch1)
+    {right_tensors, axes2, s2, names2} = prepare_batch(right, axes2, batch2)
+    {batch_shape, batch_names} = Nx.Shape.zip_reduce(s1, axes1, names1, s2, axes2, names2)
+
+    batch_out = %{out | shape: batch_shape, names: batch_names}
+
+    dotted_batches =
+      left_tensors
+      |> Stream.zip(right_tensors)
+      |> Enum.map(fn {left_batch, right_batch} ->
+        dot4(batch_out, left_batch, axes1, right_batch, axes2)
+      end)
+    concatenate(out, dotted_batches, 0)
```

Comment:
We should try to do operations on binaries as much as possible. The idea here is that the batches on both tensors will have the same dimension. So you can do something like this:

```
%{type: {_, left_size}} = left_tensor
%{type: {_, right_size}} = right_tensor

left_batch = div(Nx.size(left_tensor), batch) * left_size
right_batch = div(Nx.size(right_tensor), batch) * right_size

for i <- 0..batch-1 do
  consumed = i * left_batch
  <<_::size(consumed)-bitstring, left_inner::size(left_batch)-bitstring, _::bitstring>> = left_data
  consumed = i * right_batch
  <<_::size(consumed)-bitstring, right_inner::size(right_batch)-bitstring, _::bitstring>> = right_data
  dot(left_inner, left_axis, right_inner, right_axis)
end
```

Where the dot at the end is the dot operation that we have today. In other words, we are traversing the batch dimensions together (as if we were zipping them) and dotting them as you go. It is similar to what you are doing, but it should be more efficient because we are traversing the binary only once and not creating streams and so on.

Once you get that working, supporting multiple dimensions will be a better of changing the for comprehension to traverse on batch1*batch2 elements at the same time.  But we can cross that bridge once we get there!